Improving README.md (#29)
arduino-libraries/ArduinoIoTCloud: README.md

### Arduino IoT Cloud Components

- **Devices**: Physical objects like a hardware board (e.g. MKR WiFi 1010). This is the hardware which runs the software, read sensors, controls actuators and comunicates with the Arduino IoT Cloud.

- **Things**: Logical representation of a connected object. They represent the inherent properties of the object, with as little reference to the actual hardware used to implement them. Each thing is represented by a collection of properties (e.g., temperature, light).

- **Properties**: Qualities defining the characteristics of a system. A property can be something like a read-only (RO) setting to indicate the Arduino IoT Cloud can read the data, but cannot change the value of the property. A property might be designed as read and write (RW) if the Arduino IoT Cloud can also remotely change the property’s value and send an event notification to the device.

- **Events**: When something happens the Arduino IoT Cloud is aware of it thanks to  application messages. For example, it might be informed by a face-recognition application that someone is at a door, or it has received a request from another app that a light be turned on.

> - **Software**: It's quickly autogenerated automatically by the Arduino IoT Cloud when setting up a new thing: in this way everything that you need it's almost done and ready. The connection to the cloud it's already made possible by this library. You have only to implement the callbacks linked to the properties.

Reviewer: ```suggestion
- **Software**: It's quickly autogenerated automatically by the Arduino IoT Cloud when setting up a new thing: in this way, almost everything that you need is done and ready. The connection to the cloud is handled by this library. You have only to implement the callbacks linked to the properties.
```

## ArduinoIoTCloud library

> The library is divided in various classes:

Reviewer: ```suggestion
The library is divided into various classes:
```

> - `ConnectionManager` which is responsable for the connection to the internet and to the Arduino IoT Cloud through `WiFiConnectionManager` or `GSMConnectionManager`. The selection is done using the type of board used.

Reviewer: ```suggestion
- `ConnectionManager` which is responsable for the connection to the Internet and to the Arduino IoT Cloud through `WiFiConnectionManager` or `GSMConnectionManager`. The selection is done using the type of board used.
```

> - `ArduinoIoTCloud` it's the leading class. It's responsable for the connection to the MQTT Broker.

Reviewer: ```suggestion
- `ArduinoIoTCloud` it's the leading class. It's responsible for the connection to the MQTT Broker.
```

> - `CloudSerial` it's similar to [Serial](https://www.arduino.cc/reference/en/language/functions/communication/serial/), but it's used in combination with the cloud.

Reviewer: ```suggestion
- `CloudSerial` it's similar to [Serial](https://www.arduino.cc/reference/en/language/functions/communication/serial/), but used in combination with the cloud.
```

### Arduino IoT Cloud Components

- **Devices**: Physical objects built around a board (e.g. MKR WiFi 1010). This is the hardware which runs the sketch, reads sensors, controls actuators and communicates with the Arduino IoT Cloud.

- **Things**: Logical representation of a connected object. They embody inherent properties of the object, with as little reference to the actual hardware or code used to implement them. Each Thing is represented by a collection of _Properties_ (e.g., temperature, light, pressure...).

- **Properties**: Qualities defining the characteristics of a system. A _Property_ can be defined as *read-only* (RO) to indicate that Arduino IoT Cloud can read the data, but cannot change the value of such _Property_. On the other end it may be designated to be **read-and-write** (RW), allowing Arduino IoT Cloud to remotely change the property’s value and trigger an event notification on the device to be handled via code.

- **Events**: When a physical event is triggered on the _Device_, Arduino IoT Cloud is made aware of it thanks to application messages. It might, for example, be notified that a proximity sensor detected someone or something outside a door.

> - **Software**: An Arduino Create sketch is automatically generated by Arduino IoT Cloud when setting up a new thing: this simplifies starting efforts. Because he connection to the cloud is handled by the library, The user can focus on  implementing the last bits of code required to handle the change of variables linked to properties.

Reviewer: ```suggestion
- **Software**: An Arduino Create sketch is automatically generated by Arduino IoT Cloud when setting up a new thing: this simplifies starting efforts. Because the connection to the cloud is handled by the library, the user can focus on implementing the last bits of code required to handle the change of variables linked to properties.
```

The introductory tutorial linked above explains this in an easy and comprehensive way.

## ArduinoIoTCloud library

The library is made of multiple classes:
- `ArduinoIoTCloud` is the main class. It's responsible for the connection to the MQTT Broker and to Arduino IoT Cloud.
This library has multiple `begin(...)` methods allowing you to take more control of its behavior when it comes to network **Client** or to use a `ConnectionManager`

- `ConnectionManager` is an abstract Class defining methods to be implemented in derived classes, such as `WiFiConnectionManager`, `GSMConnectionManager` and so on. The right `ConnectionManager` is chosen on a board basis during compilation.

> - `WiFiConnectionManager` handles connection, network time retrieval, disconnection, and reconnection to Internet for WiFi equipped boards (**MKR1000**, **MKR WIFI 1010** and upcoming implementations.

Reviewer: ```suggestion
- `WiFiConnectionManager` handles connection, network time retrieval, disconnection, and reconnection to Internet for WiFi equipped boards (**MKR1000**, **MKR WIFI 1010** and upcoming implementations).
```

> - `GSMConnectionManager`. handles connection, network time retrieval, disconnection, and reconnection to Internet for GSM equipped boards (**MKR GSM 1400**)

Reviewer: ```suggestion
- `GSMConnectionManager` handles connection, network time retrieval, disconnection, and reconnection to Internet for GSM equipped boards (**MKR GSM 1400**)
```

- `CloudSerial` is similar to [Serial](https://www.arduino.cc/reference/en/language/functions/communication/serial/), but used in combination with the cloud, allowing the user to send and receive custom messages using Arduino IoT Cloud as a channel.

### ConnectionManager

**Connection Manager** is configured via a series of compiler directives, including the correct implementation of the class based on which board is selected.

### How to use it
- Instantiate the class with `ConnectionManager *ArduinoIoTPreferredConnection = new WiFiConnectionManager(SECRET_SSID, SECRET_PASS);` if you are using a WiFi board, otherwise replace **WiFi** with **GSM** or any future implementation.

> - the `check()` method does all the work. It uses a finite state machine and is responsible for connection and reconnection to a network. The method is designed to be non-blocking by using time (milliseconds) to perform its tasks.

Reviewer: ```suggestion
- The `check()` method does all the work. It uses a finite state machine and is responsible for connection and reconnection to a network. The method is designed to be non-blocking by using time (milliseconds) to perform its tasks.
```

- `getTime()` returns different implementations of the `getTIme()` method based on the board used. Time is retrieved from an NTP server and is required for the SSL connection to the cloud.

- `&getClient()` returns a reference an instance of the `Client` class used to connect to the network.

- `getStatus()` returns the network connection status. The different states are defined in an `enum`

- `debugMessage(char *_msg, uint8_t _debugLevel, bool _timestamp = true, bool _newline = true)` is the method used to print debug messages on the physical serial. This helps providing troubleshooting information should anything go wrong.

> - The `setDebugMessageLevel(int _debugLevel)` method is used to set a level of granularity in information output. Every debug message comes with a level which goes from 0 to 4. A higher level means more verbosity. Debug messages with level higher than `_debugLevel` will not been showed. The lowest level has a higher importance and is usually used for errors, which are always printed. Passing -1 as a parameter will disable logging entirely, **errors will also be ignored**.

Reviewer: ```suggestion
- The `setDebugMessageLevel(int _debugLevel)` method is used to set a level of granularity in information output. Every debug message comes with a level which goes from 0 to 4. A higher level means more verbosity. Debug messages with level higher than `_debugLevel` will not be shown. The lowest level has a higher importance and is usually used for errors, which are always printed. Passing -1 as a parameter will disable logging entirely, **errors will also be ignored**.
```

### ArduinoIoTCloud

- The `begin(ConnectionManager *connection = ArduinoIoTPreferredConnection, String brokerAddress = "mqtts-sa.iot.arduino.cc")` method is used to initialize a connection to the Arduino IoT Cloud through MQTT. Accepts an instance of `ConnectionManager` and returns 0 in case of failure or 1 if object is successfully instantiated.

- `begin(Client& net, String brokerAddress = "mqtts-sa.iot.arduino.cc")` if connection is managed manually and not via a `ConnectionManager` a reference to a `Client` object can be used.

> - The `connect()` method is used to connect to the MQTT broker

Reviewer: ```suggestion
- The `connect()` method is used to connect to the MQTT broker.
```

> - `disconnect()` closes the connection to the MQTT Client

Reviewer: ```suggestion
- `disconnect()` closes the connection to the MQTT Client.
```

- The `setDebugMessageLevel(uint8_t _debugLevel)` function is used to set a debug level. Every debug message comes with a level which goes from 0 to 4. Higher level means higher verbosity. Debug messages with level higher than `_debugLevel` will not been showed.

### ArduinoIoTCloud

- The `int begin(ConnectionManager *connection = ArduinoIoTPreferredConnection, String brokerAddress = "mqtts-sa.iot.arduino.cc")` function is used to initialize the connection to MQTT broker and to the cloud.

- The `connect()` function is used to connect to the MQTT broker and to the cloud

- `disconnect()` stops the MQTT Client

> - The `update()` function checks the connnection with the MQTT broker, retrieve data from it and pushes back data using `writeProperties()`. `update()` is called periodically in the loop of .ino file.

Reviewer 2: @umbynos I'd change the last sentence with something like "`update()` SHOULD BE called periodically in the `loop()` function."

> - `connected()` simply returns the current status of the MQTT connection.

Reviewer 2: Is this function returning just `true` or `false` or are there more statuses? I think we should write exactly what it's returning

- `reconnect(Client&)` clean up existing MQTT connection, create a new one and initialize it by calling `connect()`

> - `writeProperties(const byte data[], int length)` writes message to the MQTT Broker

Reviewer 2: @umbynos can this definition be more specific?
